<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Revse Ecology Analysis on Microbiome &bull; RevEcoR</title><!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous"><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">RevEcoR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Revse Ecology Analysis on Microbiome</h1>
            
            <h4 class="date">2016-11-21</h4>
          </div>

    
    
<div class="contents">
<div id="reverse-ecology-linking-organisms-and-their-environments" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#reverse-ecology-linking-organisms-and-their-environments" class="anchor"> </a></body></html>Reverse ecology linking organism(s) and their environments</h2>
<div id="yang-cao-yiluheiheigmail-com-fei-li-pittacusgmail-com" class="section level6">
<h6>Yang Cao <a href="mailto:yiluheihei@gmail.com">yiluheihei@gmail.com</a>, Fei Li <a href="mailto:pittacus@gmail.com">pittacus@gmail.com</a></h6>
</div>
<div id="beijing-institute-of-radiation-biotechnology-laboratory" class="section level6">
<h6>BeiJing Institute of Radiation &ndash; Biotechnology Laboratory</h6>
</div>
<div id="table-of-contens" class="section level3">
<h3 class="hasAnchor"><html><body><a href="#table-of-contens" class="anchor"> </a></body></html>Table of Contens</h3>
<ul><li><p><a href="#background">Background</a></p></li>
<li><p><a href="#intro">Introduction</a></p></li>
<li><p><a href="#install">Installation</a></p></li>
<li><p><a href="#metadata">Downloding the metabolic data</a></p></li>
<li><p><a href="#reconstruct">Reconstruction of organism metabolic network</a></p></li>
<li><p><a href="#seedset">Identify seed set of metabolic network</a></p></li>
<li><p><a href="#cooperation">Predict species interactions</a></p></li>
<li><p><a href="#microbiome">Comparing predicted interactions and co-occurrences</a></p></li>
<li><p><a href="#session">sessionInfo</a></p></li>
<li><p><a href="#ref">references</a></p></li>
</ul><p><a id="background"></a></p>
</div>
</div>
<div id="background" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#background" class="anchor"> </a></body></html>Background</h2>
<p>Ecology is the scientific With the rapid and inexpensive next-generation sequencing technologies The structure of complex biological systems reflects not only their function but also into the habitates in which they envoled and are adapted to. Reverse Ecology- an emerging new frontier in Evolutionary Sysmtems biology at the interface of computational biology, genomics and environmental science, which uses population genomics to study ecology with no a priori assumptions about the organism(s) under consideration. This term was suggested by <a href="#reversecology">Li <em>et al.</em></a> during a conference on ecological genomics in Christchurch, New Zealand. It facilitates the translation of high through genomic data into large scale ecological data, and utilizes system-based method to obtain novel insights of pooly characterized microorganisms and relationships between microorgasnims or their environments in a superorganism.Traditional approach, however, can only applied to a small scale and for ralatively well-studied systems.</p>
<p><a id="intro"></a></p>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#introduction" class="anchor"> </a></body></html>Introduction</h2>
<p>This manual is a brief introduction to structure, funcitons and usage of <strong>RevEcoR</strong> package. The <strong>RevEcoR</strong> package implements the reverse ecology framework. It can be used for reconstruction of metabolic environmnets using a cross-species analysis, idenditifying the set of compounds and predicting the species interactions on a large scale with a graph-based algorithm mentioned in <a href="#seedset">Borenstein <em>et al.</em></a>. The reverse ecology framework which takes advances of system biology and genomic metabolic modeling, aims to predict the ecological traits of poorly studied microorganisms, their interactions with other microorganisms, and the ecology of microbial communities from system-level analysis of complex biological networks.</p>
<p>Two softwares (<a href="http://elbo.gs.washington.edu/software_netseed.html">NetSeed</a> and <a href="http://elbo.gs.washington.edu/software_netcooperate.html">NetCooperate</a>), have developed by <strong>Borenstein&rsquo;s</strong> lab for studying the ecology of microorganisms. Howevever, neither of them supports metabolic network reconstruction of species, and both are limited in small scale analysis (two species). The <strong>RevEcoR</strong> package provides an interface to microbiome reverse ecology analysis on a large scale. The main features of this package consists of several steps:</p>
<ol style="list-style-type: decimal"><li><p>Downding metabolic networks data from KEGG database: all species metabolic network data could be downloaded from the KEGG PATHWAY datase with the KEGG REST API (application programming interface).</p></li>
<li><p>Reconstruction metabolic networks of all species: a directed graph whose nodes represent compounds and whose edges represent reactions linking substrates to products.</p></li>
<li><p>Identify seed set of a specific organism: each metabolic network was decomposed into its strongly conneted components (SCC) using <a href="#Kasaraju">Kasaraju&rsquo;s algorithm</a>. The SCC forms a directed graph whose nodes are the components and whose edges are the orginal edges in the graph that connect nodes in two different components. Then, detectting the seed set with this SCCs.</p></li>
<li><p>Host-microbe and microbe-microbe cooperation analysis</p></li>
</ol><p><a id="install"></a></p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#installation" class="anchor"> </a></body></html>Installation</h2>
<p><strong>RevEcoR</strong> is free available on CRAN. You can install the latest released version from CRAN as following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">"RevEcoR"</span>) </code></pre></div>
<p>or the latest development version from github. To install packages from GitHub, you first need install the <strong>devtools</strong> package on your system with <code>install.packages("devtools")</code>. Note that devtools sometimes needs some extra non-R software on your system &ndash; more specifically, an Rtools download for Windows or Xcode for OS X. There&rsquo;s more information about devtools <a href="https://github.com/hadley/devtools">here</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">if (!<span class="kw">require</span>(devtools) 
  <span class="kw">install.packages</span>(<span class="st">"devtools"</span>) 
devtools::<span class="kw">install_github</span>(<span class="st">"yiluheihei/RevEcoR"</span>) </code></pre></div>
<p>After installation, you can load <strong>RevEcoR</strong> into current workspace by typing or pasting the following codes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RevEcoR) </code></pre></div>
<p><a id="metadata"></a></p>
</div>
<div id="downloding-the-metabolic-data-of-a-specific-organism" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#downloding-the-metabolic-data-of-a-specific-organism" class="anchor"> </a></body></html>Downloding the metabolic data of a specific organism</h2>
<p>Both the Kyoto Encyclopedia of Genes and Genomes (KEGG) and Integrated Microbial Genomes database (IMG) [7] collect complete high-quality genome sequences and metagenome sequences with a comprehensive set of publicly available bacterial, archaeal, eukaryotic, and phage genomes, as well as engineered, environmental and host associated metagenome samples. All the sequences and functional annotation profile can be download directly.</p>
<p>The KEGG database provides a REST-style API compared with IMG database. This package provides function <code>getOrgMetabolicData</code> to download the specific organism metabolic data from KEGG database, and return a list where each element consists of three elements: reaction, substrate and product. Data in IMG, we can only download it manually. In addition, users can annotate their private genomic data with KO terms on IMG systems or in their local machine to obtain the annotation profile.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## download sample metabolic data from remote KEGG database 
buc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getOrgMetabolicData.html">getOrgMetabolicData</a></span>(<span class="st">"buc"</span>) 
<span class="kw">data</span>(kegg_buc) 
<span class="kw">head</span>(buc) </code></pre></div>
<p><a id="reconstruct"></a> ## Reconstruction of organism metabolic network</p>
<p>The crucial premise of Reverse Ecology is that genomic information can be converted into ecological information. Metabolism linking miroorgasnim with the biochemical environmnet acrossing compound exchanges: import of exogenous compounds and impact the composition of its environment via secretion of other compounds. Thus, metabolic network is the reflection of interation between organisms and their environment.</p>
<p>Graph-based representation of metabolic reactions where nodes represent compounds and edges represent reactions is a common tool in analyzing and studing metabolic networks. A directed edge from node a to b indicates that compound a is a substrate in some reaction that produces compound b.</p>
<p>Once the metabolic data is obtained, <code>reconstructGsMN</code> could be used to reconstruction the metabolic network of a specific organism.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## species in KEGG 
buc.net &lt;-<span class="st"> </span><span class="kw"><a href="../reference/reconstructGsMN.html">reconstructGsMN</a></span>(kegg_buc, <span class="dt">RefData =</span> <span class="ot">NULL</span>) 
igraph::<span class="kw">print.igraph</span>(buc.net) </code></pre></div>
<pre><code>## IGRAPH DN-- 259 681 -- 
## + attr: name (v/c)
## + edges (vertex names):
##  [1] cpd:C00022-&gt;cpd:C00002 cpd:C00022-&gt;cpd:C00008 cpd:C00022-&gt;cpd:C00074
##  [4] cpd:C00022-&gt;cpd:C00024 cpd:C00022-&gt;cpd:C06010 cpd:C00022-&gt;cpd:C00109
##  [7] cpd:C00022-&gt;cpd:C00118 cpd:C00022-&gt;cpd:C00900 cpd:C00022-&gt;cpd:C05125
## [10] cpd:C00022-&gt;cpd:C06006 cpd:C00022-&gt;cpd:C11437 cpd:C00068-&gt;cpd:C01081
## [13] cpd:C00068-&gt;cpd:C05125 cpd:C00068-&gt;cpd:C05381 cpd:C04332-&gt;cpd:C00255
## [16] cpd:C04332-&gt;cpd:C04732 cpd:C04332-&gt;cpd:C15556 cpd:C00255-&gt;cpd:C04332
## [19] cpd:C00255-&gt;cpd:C00061 cpd:C00255-&gt;cpd:C04732 cpd:C00003-&gt;cpd:C00006
## [22] cpd:C00003-&gt;cpd:C00455 cpd:C00003-&gt;cpd:C00857 cpd:C00006-&gt;cpd:C00003
## + ... omitted several edges</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">igraph::<span class="kw">plot.igraph</span>(buc.net, <span class="dt">vertex.label=</span><span class="ot">NA</span>, <span class="dt">vertex.size=</span><span class="dv">5</span>, <span class="dt">edge.arrow.size=</span><span class="fl">0.1</span>)</code></pre></div>
<p><img src="RevEcoR_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## ko annotation profile species detected in a human microbiome in IMG (not in KEGG) 
annodir &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/koanno.tab"</span>,<span class="dt">package =</span> <span class="st">"RevEcoR"</span>) 
metabolic.data &lt;-<span class="st"> </span><span class="kw">read.delim</span>(annodir,<span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>) 
##load the reference metabolic data 
<span class="kw">data</span>(RefDbcache) 
g2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/reconstructGsMN.html">reconstructGsMN</a></span>(metabolic.data, <span class="dt">RefData =</span> RefDbcache) </code></pre></div>
<p><a id="seedset"></a></p>
</div>
<div id="identify-seed-set-of-metabolic-network" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#identify-seed-set-of-metabolic-network" class="anchor"> </a></body></html>Identify seed set of metabolic network</h2>
<p>As the interactions with the environment was reflected in the metabolic networks, these networks could be used not only to infer metabolic function but alse to obtain insights into the growth environments in which the species evolved.</p>
<p>Apparantly, organisms can survive in a wide range of enviromnets and may activate only a subset of the pathways in the network of each environment, using a different set of exogenously acquired compounds (termed seed set). The seed set of the network is defined as the minimal set of compounds in the network that allows the synthesis of all other compouds, and can serve as a good proxy for its environment and can be conceived as the essential and effective biochemical environment.</p>
<p><code>getSeedSets</code> was used to detect the seed set of a metabolic network which returns a seedset-class object. Futhermore, some methods was supported for seedset-class, e.g length, nonseeds. For more details on seedset-class, see <code>?seedset-class</code>.It can help us to get the compound that organisms are exogenously acquired compouds from the environment, representing the organism&rsquo;s nutritional profile. This algorithm is based on a fast method Kasaraju algorithm for SCC decomposition which is implementaed in <code>Kasaraju</code>. For more details, see <code><a href="../reference/getSeedSets.html">?getSeedSets</a></code> and <code>?KasarajuSCC</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## seed set prediction
seed.set &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getSeedSets.html">getSeedSets</a></span>(buc.net, <span class="fl">0.2</span>) 
<span class="kw"><a href="../reference/show,seedset-method.html">show</a></span>(seed.set) </code></pre></div>
<pre><code>## Object of class seedset
##    IGRAPH: -- 259 nodes 681 edges -- 
##   seedset length  9</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(seed.set@seeds)</code></pre></div>
<pre><code>## [[1]]
## [1] "cpd:C15532"
## 
## [[2]]
## [1] "cpd:C15811"
## 
## [[3]]
## [1] "cpd:C05729"
## 
## [[4]]
## [1] "cpd:C00993"
## 
## [[5]]
## [1] "cpd:C00031"
## 
## [[6]]
## [1] "cpd:C00392"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## The node colored with red represents the species' seed set
nodes  &lt;-<span class="st"> </span>igraph::<span class="kw">V</span>(buc.net)$name
seeds  &lt;-<span class="st"> </span><span class="kw">unlist</span>(seed.set@seeds)
seed.index  &lt;-<span class="st"> </span><span class="kw">match</span>(seeds,nodes)
node.color &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="st">"SkyBlue2"</span>,<span class="kw">length</span>(nodes))
node.color[seed.index]  &lt;-<span class="st"> "red"</span>
igraph::<span class="kw">plot.igraph</span>(buc.net, 
          <span class="dt">vertex.label=</span><span class="ot">NA</span>, <span class="dt">vertex.size=</span><span class="dv">5</span>, <span class="dt">edge.arrow.size=</span><span class="fl">0.1</span>,
          <span class="dt">vertex.color =</span> node.color)</code></pre></div>
<p><img src="RevEcoR_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<p><a id="cooperation"></a></p>
</div>
<div id="predict-species-interactions" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#predict-species-interactions" class="anchor"> </a></body></html>Predict species interactions</h2>
<p>The topology of metabolic networks can provide insight not only into the metabolic process that accur within each species, but also into interactions between different species. Here we provides <code>calculateCoopreationIndex</code> using three cooperation index: <a href="#rule">competition index, coplementarity index</a> measure the microbe-microbe co-occurrence pattern. More details, see <code>?calculateCoopreationIndex</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># ptr metabolic network </span>
<span class="kw">data</span>(kegg_ptr) 
##ptr.net &lt;- reconstructGsMN(getOrgMetabolicData("ptr")) 
ptr.net &lt;-<span class="st"> </span><span class="kw"><a href="../reference/reconstructGsMN.html">reconstructGsMN</a></span>(kegg_ptr) 
<span class="co"># cooperation analysis between buc and ptr </span>
cooperation.index &lt;-<span class="st"> </span><span class="kw"><a href="../reference/calculateCooperationIndex.html">calculateCooperationIndex</a></span>(buc.net,ptr.net) 
cooperation.index </code></pre></div>
<pre><code>## $competition.index
##            [,1]      [,2]
## [1,] 1.00000000 0.1818182
## [2,] 0.03571429 1.0000000
## 
## $complementarity.index
##            [,1]      [,2]
## [1,] 0.00000000 0.4545455
## [2,] 0.07142857 0.0000000</code></pre>
<p>Blow we applied <strong>RevEcoR</strong> to predict interactions among several human oral microbiota species whose co-occurrence patterns have previously been described. Our seven sample oral species metabolic dataset was downloaded from the IMG server; it consists of the following <a href="http://www.genomesonline.org">Genomes Online Database (GOD)</a> IMG identification numbers: Gc0016386, Gi07614, Gi00264, Gc00809, Gc00643, Gi03876, and Gi07289. The interactions among the seven species is calculated as follwing:</p>
<p>The metabolic networks of seven species was reconstructed first:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">##metabolic network reconstruction of these seven species
net  &lt;-<span class="st"> </span><span class="kw">lapply</span>(anno.species, reconstructGsMN)</code></pre></div>
<p>Then <code>caculateCooperationIndex</code> is used to calculated the species interactions among the seven species.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## caculate interactions among vious species
interactions  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/calculateCooperationIndex.html">calculateCooperationIndex</a></span>(net, <span class="dt">p =</span> <span class="ot">TRUE</span>)
## competition index
$competition.index
          Aa        Ao        Fn        Pg        Sg        So        Va
Aa <span class="fl">1.0000000</span> <span class="fl">0.4736842</span> <span class="fl">0.3157895</span> <span class="fl">0.2280702</span> <span class="fl">0.4210526</span> <span class="fl">0.4385965</span> <span class="fl">0.2456140</span>
Ao <span class="fl">0.4736842</span> <span class="fl">1.0000000</span> <span class="fl">0.3684211</span> <span class="fl">0.3333333</span> <span class="fl">0.4736842</span> <span class="fl">0.4736842</span> <span class="fl">0.2456140</span>
Fn <span class="fl">0.5000000</span> <span class="fl">0.5833333</span> <span class="fl">1.0000000</span> <span class="fl">0.4166667</span> <span class="fl">0.5833333</span> <span class="fl">0.5555556</span> <span class="fl">0.4166667</span>
Pg <span class="fl">0.4193548</span> <span class="fl">0.6129032</span> <span class="fl">0.4838710</span> <span class="fl">1.0000000</span> <span class="fl">0.6129032</span> <span class="fl">0.5161290</span> <span class="fl">0.3870968</span>
Sg <span class="fl">0.5454545</span> <span class="fl">0.6136364</span> <span class="fl">0.4772727</span> <span class="fl">0.4318182</span> <span class="fl">1.0000000</span> <span class="fl">0.9090909</span> <span class="fl">0.3863636</span>
So <span class="fl">0.5813953</span> <span class="fl">0.6046512</span> <span class="fl">0.4651163</span> <span class="fl">0.3720930</span> <span class="fl">0.9302326</span> <span class="fl">1.0000000</span> <span class="fl">0.3953488</span>
Va <span class="fl">0.4827586</span> <span class="fl">0.4827586</span> <span class="fl">0.5172414</span> <span class="fl">0.4137931</span> <span class="fl">0.5862069</span> <span class="fl">0.5862069</span> <span class="fl">1.0000000</span>
## p value of competition index
$competition.index.p
Aa    Ao    Fn    Pg    Sg    So    Va
Aa <span class="fl">0.000</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span>
Ao <span class="fl">0.001</span> <span class="fl">0.000</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span>
Fn <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.000</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span>
Pg <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.000</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span>
Sg <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.000</span> <span class="fl">0.001</span> <span class="fl">0.001</span>
So <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.000</span> <span class="fl">0.001</span>
Va <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.000</span>
## complementarity index
$complementarity.index
 Aa        Ao        Fn         Pg        Sg         So        Va
Aa <span class="fl">0.0000000</span> <span class="fl">0.1052632</span> <span class="fl">0.1228070</span> <span class="fl">0.07017544</span> <span class="fl">0.0877193</span> <span class="fl">0.08771930</span> <span class="fl">0.1228070</span>
Ao <span class="fl">0.1403509</span> <span class="fl">0.0000000</span> <span class="fl">0.1403509</span> <span class="fl">0.07017544</span> <span class="fl">0.1228070</span> <span class="fl">0.12280702</span> <span class="fl">0.1403509</span>
Fn <span class="fl">0.1944444</span> <span class="fl">0.1666667</span> <span class="fl">0.0000000</span> <span class="fl">0.16666667</span> <span class="fl">0.1111111</span> <span class="fl">0.11111111</span> <span class="fl">0.1388889</span>
Pg <span class="fl">0.2258065</span> <span class="fl">0.2258065</span> <span class="fl">0.1612903</span> <span class="fl">0.00000000</span> <span class="fl">0.1612903</span> <span class="fl">0.19354839</span> <span class="fl">0.2258065</span>
Sg <span class="fl">0.2272727</span> <span class="fl">0.1818182</span> <span class="fl">0.1590909</span> <span class="fl">0.09090909</span> <span class="fl">0.0000000</span> <span class="fl">0.04545455</span> <span class="fl">0.1590909</span>
So <span class="fl">0.1860465</span> <span class="fl">0.1395349</span> <span class="fl">0.1860465</span> <span class="fl">0.09302326</span> <span class="fl">0.0000000</span> <span class="fl">0.00000000</span> <span class="fl">0.1395349</span>
Va <span class="fl">0.2068966</span> <span class="fl">0.1724138</span> <span class="fl">0.1379310</span> <span class="fl">0.17241379</span> <span class="fl">0.1379310</span> <span class="fl">0.13793103</span> <span class="fl">0.0000000</span>
## p value of complementarity index
$complementarity.index.p
 Aa    Ao    Fn    Pg    Sg    So    Va
Aa <span class="fl">0.000</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span>
Ao <span class="fl">0.001</span> <span class="fl">0.000</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span>
Fn <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.000</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span>
Pg <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.000</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span>
Sg <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.000</span> <span class="fl">0.001</span> <span class="fl">0.001</span>
So <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.000</span> <span class="fl">0.001</span>
Va <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.001</span> <span class="fl">0.000</span></code></pre></div>
<p>We found that Streptococcus oralis (So) and Streptococcus gordonii (Sg) had the lowest complementarity index (0.04 and 0.00, respectively) and the highest competition index (0.91 and 0.93, respectively) among all pairs. Appanrently, the p value of all interactions are biologcial/statistical significance. This indicates that these two species are antagonistic, which is consistent with previous findings.</p>
<p><a id="microbiome"></a></p>
</div>
<div id="comparing-predicted-interactions-and-co-occurrences" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#comparing-predicted-interactions-and-co-occurrences" class="anchor"> </a></body></html>Comparing predicted interactions and co-occurrences</h2>
<p>To further evaluate the predicted interactions of RevEcoR on a large scale, we integrated functions mentioned above to investigate species interactions in the gut microbiome. We focused on a list of 116 prevalent gut species, whose genome sequence is available in IMG database and sequence coverage is more than 1% in at least one metagenomic sample of <a href="#gut">124 individuals</a>. Genome annotation profiles of this 116 species was collected from IMG database and was used to calculated the interactions (competition and complementarity index) for all pairs of species.</p>
<div id="predicting-species-interactions-in-gut-microbiome" class="section level3">
<h3 class="hasAnchor"><html><body><a href="#predicting-species-interactions-in-gut-microbiome" class="anchor"> </a></body></html>1. Predicting species interactions in gut microbiome</h3>
<p>For each species, we download the list of genes mapped to the Kyoto Encyclopedia of Genes and Genomes orthologous groups (KOs) was downloaded with a in-house R script. This data, which was used to reconstruct the metabolic network of each species, have been saved as <em>gut_microbiome.rda</em> in subdirectory <em>data</em> of RevEcoR. We can load it as the following code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(gut_microbiome) 
## summary(gut_microbiome) </code></pre></div>
<p>Then, it was used to reconstuct the metabolic network, predict the seed set, and finally predict the pairs of interactions between different species:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gut.nets &lt;-<span class="st"> </span><span class="kw">lapply</span>(gut_microbiome,reconstructGsMN)
seed.sets &lt;-<span class="st"> </span><span class="kw">lapply</span>(gut.nets,getSeedSets)
## Since calculation is on large scale, species interactions prediction may take several hours
gut.interactions &lt;-<span class="st"> </span><span class="kw"><a href="../reference/calculateCooperationIndex.html">calculateCooperationIndex</a></span>(gut.nets)
competition.index &lt;-<span class="st"> </span>gut.interactions$competition.index
complementarity.index &lt;-<span class="st"> </span>gut.interactions$complementarity.index</code></pre></div>
</div>
<div id="obtaining-co-occurrence-scores-in-gut-microbiome" class="section level3">
<h3 class="hasAnchor"><html><body><a href="#obtaining-co-occurrence-scores-in-gut-microbiome" class="anchor"> </a></body></html>2. Obtaining co-occurrence scores in gut microbiome</h3>
<p>Specially, it will help us to predict whether species compete with one another tend to co-occur or to exclude by comparing the species interactions and co-occurrences. We obtained co-occurrence scores directly from <a href="#rule"><em>Levy R&rsquo;s</em> research</a>, and saved it as <em>occurence.tab</em> in subdirectory <em>inst/extdata</em> of RevEcoR . Co-occurrence score was calculated based on species abundances across all samples and measured by the Jaccard similarity index.</p>
<p>Load the co-occurrence data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">occurrence.score &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="kw">system.file</span>(<span class="st">"extdata/occurrence.tab"</span>,
  <span class="dt">package =</span> <span class="st">"RevEcoR"</span>),<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>, <span class="dt">quote =</span> <span class="st">""</span>)</code></pre></div>
</div>
<div id="comparing-the-interactions-and-co-occurrences" class="section level3">
<h3 class="hasAnchor"><html><body><a href="#comparing-the-interactions-and-co-occurrences" class="anchor"> </a></body></html>3. Comparing the interactions and co-occurrences</h3>
<p>The Jaccard index measures similarity between finite sample sets, and is defined as the size of the intersection divided by the size of the union of the sample sets. Thus, co-occurrence scores are generally symmetric whereas competition index or complementarity index are not according to their definitions.</p>
<p>A symmetric version of two interaction index was gnerated by replacing each element of the interaction indices with the mean of each value and that transpose value.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">competition.index &lt;-<span class="st"> </span>(competition.index +<span class="st"> </span><span class="kw">t</span>(competition.index))/<span class="dv">2</span>
complementarity.index &lt;-<span class="st"> </span>(complementarity.index +<span class="st"> </span><span class="kw">t</span>(complementarity.index))/<span class="dv">2</span></code></pre></div>
<p>Subsequently, the Spearman correlation between the co-occurrence scores and the two interaction indices was calculated. A permutation-based Mantel test, which is commonly used in ecology, was used to determin the significance of this correlation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## upper triangles, which is used to calculate the correlation
competition.upper &lt;-<span class="st"> </span>competition.index[<span class="kw">upper.tri</span>(competition.index)]
occurrence.upper &lt;-<span class="st"> </span>occurrence.score[<span class="kw">upper.tri</span>(occurrence.score)]
complementarity.upper &lt;-<span class="st"> </span>complementarity.index[<span class="kw">upper.tri</span>(complementarity.index)]

## calculate the spearman correlation betwwen co-occurrence scores and two 
## interactions indices
competition.cor &lt;-<span class="st"> </span><span class="kw">cor</span>(competition.upper,occurrence.upper,<span class="dt">method=</span><span class="st">"spearman"</span>)
complementarity.cor &lt;-<span class="st"> </span><span class="kw">cor</span>(complementarity.upper,occurrence.upper,<span class="dt">method=</span><span class="st">"spearman"</span>)

## permutation-based mantel test. Random permutation the co-occurance score 
## 10000 times, P value is the fraction of correlations as high as or higher 
## than the original
if (<span class="kw">require</span>(magrittr)){
  null.stat &lt;-<span class="st"> </span><span class="kw">replicate</span>(<span class="dv">10000</span>,
    <span class="kw">sample</span>(<span class="dv">1</span>:<span class="dv">116</span>) %&gt;%<span class="st"> </span>occurrence.score[.,.] %&gt;%
<span class="st">      </span>.[<span class="kw">upper.tri</span>(.)]
  )
  competition.null &lt;-<span class="st"> </span><span class="kw">cor</span>(competition.upper,null.stat)
  complementarity.null &lt;-<span class="st"> </span><span class="kw">cor</span>(complementarity.upper,null.stat)
  <span class="kw">length</span>(<span class="kw">which</span>(competition.null &gt;=<span class="st"> </span>competition.cor)) ## 0 p.competition &lt; 0.00001
  <span class="kw">length</span>(<span class="kw">which</span>(complementarity.null &lt;=<span class="st"> </span>complementarity.cor)) ## 0 p.complementarity&lt; 0.00001
}</code></pre></div>
<p>We found that competition index is significant positively correlated with co-occurrence (cor = 0.261, P &lt; 10-4, Mantel correlation test), whereas the complementarity index is significant negatively correlated with co-occurrence (cor = -0.259, P &lt; 10-4, Mantel correlation test). This suggests that competition is liked to be the key factor to promote the assembly of gut microorganisms</p>
<p><a id="session"></a></p>
</div>
</div>
<div id="sessioninfo" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#sessioninfo" class="anchor"> </a></body></html>sessionInfo</h2>
<p>The version number of R and packages loaded for generating the vignette were:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>() </code></pre></div>
<pre><code>## R version 3.3.2 (2016-10-31)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: macOS Sierra 10.12.1
## 
## locale:
## [1] zh_CN.UTF-8/zh_CN.UTF-8/zh_CN.UTF-8/C/zh_CN.UTF-8/zh_CN.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] knitr_1.15           RevEcoR_0.99.3       pkgdown_0.1.0.9000  
## [4] devtools_1.12.0.9000
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.8         git2r_0.15.0        plyr_1.8.3         
##  [4] tools_3.3.2         highlight_0.4.7     digest_0.6.10      
##  [7] pkgbuild_0.0.0.9000 pkgload_0.0.0.9000  memoise_1.0.0      
## [10] evaluate_0.10       tibble_1.2-12       lattice_0.20-34    
## [13] Matrix_1.2-7.1      igraph_1.0.1        curl_2.1           
## [16] yaml_2.1.14         withr_1.0.2         httr_1.2.1         
## [19] stringr_1.1.0.9000  roxygen2_5.0.1      xml2_1.0.0.9001    
## [22] desc_1.0.1          gtools_3.5.0        rprojroot_1.1      
## [25] grid_3.3.2          R6_2.2.0            XML_3.98-1.4       
## [28] rmarkdown_1.1.9017  callr_1.0.0.9000    purrr_0.2.2.9000   
## [31] magrittr_1.5        whisker_0.3-2       backports_1.0.4    
## [34] htmltools_0.3.5     assertthat_0.1      stringi_1.1.2      
## [37] lazyeval_0.2.0.9000 crayon_1.3.2</code></pre>
<p><a id="ref"></a></p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#references" class="anchor"> </a></body></html>references</h2>
<ul><li><p><a name="reversecology">Li Y F, Costello J C, Holloway A K, et al. &ldquo;Reverse ecology&rdquo; and the power of population genomics[J]. Evolution, 2008, 62(12): 2984-2994.</a></p></li>
<li><p><a name="seedset">Borenstein E, Kupiec M, Feldman M W, et al. Large-scale reconstruction and phylogenetic analysis of metabolic environments[J]. Proceedings of the National Academy of Sciences, 2008, 105(38): 14482-14487.</a></p></li>
<li><p><a name="Kasaraju">Tarjan R. Depth-first search and linear graph algorithms[J]. SIAM journal on computing, 1972, 1(2): 146-160.</a></p></li>
<li><p><a name="rule">Levy R, Borenstein E. Metabolic modeling of species interaction in the human microbiome elucidates community-level assembly rules[J]. Proceedings of the National Academy of Sciences, 2013, 110(31): 12804-12809.</a></p></li>
<li><p><a name="bsi">Borenstein E, Feldman M W. Topological signatures of species interactions in metabolic networks[J]. Journal of Computational Biology, 2009, 16(2): 191-200.</a></p></li>
<li><p><a name="gut">Qin J, Li R, Raes J, et al. A human gut microbial gene catalogue established by metagenomic sequencing[J]. Nature, 2010, 464(7285): 59-65.</a></p></li>
</ul></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked"><li><a href="#reverse-ecology-linking-organisms-and-their-environments">Reverse ecology linking organism(s) and their environments</a></li>
      <li><a href="#background">Background</a></li>
      <li><a href="#introduction">Introduction</a></li>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#downloding-the-metabolic-data-of-a-specific-organism">Downloding the metabolic data of a specific organism</a></li>
      <li><a href="#identify-seed-set-of-metabolic-network">Identify seed set of metabolic network</a></li>
      <li><a href="#predict-species-interactions">Predict species interactions</a></li>
      <li><a href="#comparing-predicted-interactions-and-co-occurrences">Comparing predicted interactions and co-occurrences</a></li>
      <li><a href="#sessioninfo">sessionInfo</a></li>
      <li><a href="#references">references</a></li>
      </ul></div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://choyang.me">Yang Cao</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer></div>

  </body></html>
